include/mbedtls/config.h

```diff
@@ -1326,6 +1326,22 @@
  */
 #define MBEDTLS_PKCS1_V21
 
+/** \def MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS
+ *
+ * Enable support for platform built-in keys. If you enable this feature,
+ * you must implement the function mbedtls_psa_platform_get_builtin_key().
```

Comment:
I'd like to get an assertion statement in here that the platform function is meant to be autogenerated in the future, as agreed.

Reply:
The platform function is not meant to be autogenerated. It can also declare non-secure-element keys, and the assignment of key ids is not always automatic.

I'd like to simplify the work for integrators who want to just plug in a driver with a minimum of fuss, but it isn't as simple as making this function autogenerated.

Reply:
On Thursday, it was still going to be autogenerated:
> In the future, this platform callback is intended to be generated from JSON that declares the built-in keys of each driver. There are many details to work out, which is why for the time being platform vendors will hardcode what they need in this callback.

What changed?

Reply:
I realized that it can't be generated solely from driver descriptions, since keys derived from a HUK would typically not involve a driver.

The answer may be that HUK derivation should be covered by a driver that only provides a key derivation interface. Or that the autogeneration needs an extra hook. Or that there should be an intermediate interface. Or something else. I'm still not sure what the right thing is at this stage.

Reply:
keys derived from a HUK wouldn't strictly be builtin keys right? They'd be keys derived from a builtin key. And for that, there's the PSA key derivation interface...

Reply:
From the point of view of an application, there's typically a “master” device-bound key. Whether this key is stored directly in trusted storage, or derived from a key that's stored in a special location (e.g. fuses), or resides in a secure element, is an implementation detail that the platform code should hide. If the key is derived from a HUK, the application doesn't see the HUK, and in particular can't derive arbitrary keys from the HUK: it only has access to one specific key that's derived from the HUK, the application identity and possibly the device state.

Reply:
At the end of the day, in order to access any key, there needs to be both a way to know that the key is there, and a way to either access the key or exercise the key. This can be solved by the core using autogeneration for keys exposed by drivers. For derived keys that drivers aren't involved in, such as potentially the use case you're describing, you'd need a provisioning and access wrapper regardless (which IMO is not part of PSA Crypto's mission to provide).

There can be autogeneration (and I think there should be) for key access that can be automated. That means keys defined by drivers, potentially mixed with information from application- (or platform-)provided JSON (which is going to happen anyway in order for the application to make sure its algorithmic requirements are provided for) leads to a hardcoded key ID to location/slot map. Potentially with hooks for some type of access control mechanism. That should be enough to set up the mapping to the keys that the hardware driver knows about. The rest is IMO platform-specific logic that would be so broad that there won't be a good way to solve all use cases. For platform-derived keys that don't have anything to do with the hardware, I'd rather think about a different mechanism, since it's a different problem.